File: Array2.java
//Time Complexity O(N) and space O(1)
class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {


        //mark the numbers as negative
        for(int i=0;i<nums.length;i++){        
            int l = Math.abs(nums[i])-1;
            if(nums[l]>0){
                nums[l]*=-1;
            }
        }

        List<Integer> list = new ArrayList<>();
        for(int i=1;i<=nums.length;i++){
            if(nums[i-1]>0){
                list.add(i);
            }
        }

        return list;
        
    }
}

//Game of life
//time complexity O(M X N) and space complexity O(1)
class Solution {

    int[][] dirs;

    public void gameOfLife(int[][] board) {

        //make direction array
        dirs = new int[][]{{-1,0},{1,0},{0,-1},{0,1},{-1,-1},{-1,1},{1,-1},{1,1}};
        int m = board.length;
        int n = board[0].length;

        //converting the array into -1 and 2
                            for(int i=0;i<m;i++){
                                for(int j=0;j<n;j++){

                                    int countlive = countlive(board,i,j);
                                    if(board[i][j]==1 && (countlive<2 || countlive >3)){
                                        board[i][j] =-1;
                                    }else if(board[i][j]==0 && countlive==3){
                                        board[i][j]=2;
                                    }
                                }
                            }

    // re-converting the array back to new gen
                                for(int i = 0;i<m;i++){
                                    for (int j = 0; j < n; j++){
                                        if (board[i][j] == -1) {
                                            board[i][j] = 0;
                                        } else if (board[i][j] == 2) {
                                            board[i][j] = 1;
                                        }
                                    }
                                }

    }

    // function to count live cells

    public int countlive(int[][]board,int r,int c){
        int count =0;
        for(int[] dir:dirs){
            int nr = dir[0] + r;
            int nc = dir[1] + c;

            if(nr>=0 && nc>=0 && nr<board.length && nc<board[0].length&& (board[nr][nc]==-1|| board[nr][nc]==1) ){
                count++;
            }
        }

        return count;
        
    }

}

File: Array_2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


namespace S30_Problems
{
    class Program
    {
        static void Main(string[] args)
        {
            Program p = new Program();
            p.GameOfLife(new int[][] { new int[] { 0,1, 0}, new int[] { 0, 0, 1 }, new int[] { 1,1,1 }, new int[] { 0, 0, 0 } });
        }
        // Time Complexity: O(2n)
        // Space Complexity: O(1)
        public IList<int> FindDisappearedNumbers(int[] nums) {
            IList<int> res = new List<int>();
            for(int i = 0; i < nums.Length; i++){
                if(nums[Math.Abs(nums[i])-1]>0){
                    nums[Math.Abs(nums[i])-1] *= -1; 
                }
            }
            for(int i = 0; i < nums.Length; i++){
                if(nums[i]>0) res.Add(i+1);
                else nums[i]*=-1;
            }
            return res;
        }

        // Time Complexity: O(m*n*9) 8 * m*n for calculates 8 directions for each cell + 1 * m*n for rewriting the board.
        // Space Complexity: O(1)

        int[][] dirs = new int[][] {new int[]{0,1},new int[]{1,0},new int[]{0,-1},new int[]{-1,0},new int[]{-1,-1},new int[]{-1,1},new int[]{1,-1},new int[]{1,1}};
        int[][] board;
        public void GameOfLife(int[][] board) {
            this.board = board;
            for(int i = 0; i<board.Length; i++){
                for(int j = 0; j<board[0].Length; j++){
                    int count = countAlive(i,j);
                    if(board[i][j] == 0){
                        if(count == 3) board[i][j] = -2;
                    }
                    else{
                        if(count < 2 || count > 3) board[i][j] = -1;
                    }
                }
            }
            
            for(int i = 0; i<board.Length; i++){
                for(int j = 0; j<board[0].Length; j++){
                    if(board[i][j] == -2) board[i][j] = 1;
                    if(board[i][j] == -1) board[i][j] = 0;
                }
            }
        }
        
        private int countAlive(int i, int j){
            int count = 0;
            foreach(int[] dir in dirs){
                int ni = i+dir[0];
                int nj = j+dir[1];
                
                if(ni<0 || nj < 0 || ni>=board.Length || nj >= board[0].Length ) continue;
                if(board[ni][nj] == 1 || board[ni][nj] == -1) count++;
            }
            return count;
        }
    }
}
